src/bitmap/backend/atomic_bitmap_mmap.rs

```diff
@@ -0,0 +1,608 @@
+
+// SAFETY: Sync is not automatically implemented because the raw pointer.
+// `MmapLogReg` doesn't have any interior mutability and all access to `&AtomicU8`
+// are done through atomic operations.
+unsafe impl Sync for MmapLogReg {}
+
+impl MmapLogReg {
+    fn from_file(fd: BorrowedFd, offset: isize, len: usize) -> io::Result<Self> {
+        #[cfg(not(miri))]
+        // SAFETY: `fd` is a valid file descriptor and we are not using `libc::MAP_FIXED`.
+        let addr = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                len as libc::size_t,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset as libc::off_t,
+            )
+        };
+
+        #[cfg(not(miri))]
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        #[cfg(miri)]
+        if len == 0 {
+            return Err(io::Error::from_raw_os_error(libc::EINVAL));
+        }
+
+        // Miri does not support the mmap syscall, so we use rust's allocator for miri tests
+        #[cfg(miri)]
+        let addr = unsafe {
+            std::alloc::alloc_zeroed(std::alloc::Layout::from_size_align(len, 1).unwrap())
+        };
+
+        Ok(Self {
+            addr: addr as *const AtomicU8,
+            len,
+        })
+    }
+
+    fn base_ptr(&self) -> *const AtomicU8 {
+        self.addr
+    }
+}
+
+impl Index<usize> for MmapLogReg {
+    type Output = AtomicU8;
+
+    // It's ok to get a reference to an atomic value.
+    fn index(&self, index: usize) -> &Self::Output {
+        assert!(index < self.len);
+        // Note: Instaed of `&*` we can use `AtomicU8::from_ptr()` as soon it gets stabilized.
+        // SAFETY: `self.addr` + `index` doesn't wrap around and is contained within the mapped
+        // memory region.
+        unsafe { &*self.addr.add(index) }
+    }
+}
+
+impl Drop for MmapLogReg {
+    fn drop(&mut self) {
+        // SAFETY: `addr` is non-null and properly aligned, also we are sure that this
+        // is the last reference alive and/or we have an exclusive access to this object.
+        unsafe {
+            #[cfg(not(miri))]
+            libc::munmap(self.addr as *mut libc::c_void, self.len as libc::size_t);
+
+            #[cfg(miri)]
+            std::alloc::dealloc(
+                self.addr as *mut u8,
+                std::alloc::Layout::from_size_align(self.len, 1).unwrap(),
+            );
+        }
+    }
+}
+
+#[inline]
+fn try_into<TI: TryInto<TO>, TO>(val: TI) -> io::Result<TO> {
+    val.try_into()
+        .map_err(|_| io::Error::from(io::ErrorKind::InvalidData))
+}
+
+#[inline]
+fn check_pointer_offset(ptr: *const AtomicU8, offset: usize) -> io::Result<()> {
+    let offset: isize = try_into(offset)?;
+    if ptr as usize > (isize::MAX - offset) as usize {
+        return Err(io::Error::from(io::ErrorKind::InvalidData));
+    }
```

Comment:
This forces `ptr as usize <= isize::MAX as usize`. On x86-64 where the kernel decides to use the higher half, that’s generally true. What about other platforms? I know for sure that on 32-bit x86 some kernels decided to only start at 0xc0000000 or 0xf0000000.

Reply:
This is a requirement of std::ptr add() method.
I know that can be a problem in 32bit with PAE, but in that case some part of rust will not work either, like vectors.

Reply:
As far as I understand, the only requirement of the `add()` method is that the offset will not overflow `isize`, i.e. that for a pointer `*T`, `count * size_of::<T>() <= isize::MAX`. However, it does not seem to place that limitation on the pointer (neither `self` nor the result), rather the opposite, it says that “the infinite-precision sum must fit in a `usize`.”

However, this code here checks that the pointer itself will not overflow `isize`, which I think is wrong. Pointers (in integer representation) outgrowing `isize` has nothing to do with PAE, as I’ve said, I believe on x86, addresses exceeding 0x80000000 were not entirely unusual even in userspace.